openclsim/plugins/weather.py

```diff
@@ -196,3 +208,345 @@ def process_data(self, criterion) -> None:
         }
         return result
+
+
+# -------------------------------------------------------------------------------------!
+class HasOperationalLimits(object):
+    """
+    Provide operational constraints to an activity or equipment.
+
+    The HasOperationLimits class allows the user to set operational
+    limits to an activity or equipment. Though, the class is mainly
+    setup to be inherited from by other classes, such as the
+    `HasRequestWindowPluginActivity` (see below).
+
+    Parameters
+    ----------
+        limit_expr: Callable
+            A (Python) function expressing the operational limits in
+            terms of critical parameters, for example, the significant
+            wave height and peak wave period `f(hs, tp)`. The function
+            should return a bool, where `True` is considered as the
+            event in which the limit has been exceeded.
+
+    """
+
+    def __init__(self, limit_expr: Callable, *args, **kwargs):
+        """Class constructor."""
+        super().__init__(*args, **kwargs)
+
+        # Instance attributes
+        self.limit_expr = limit_expr
+
+
+# -------------------------------------------------------------------------------------!
+class HasRequestWindowPluginActivity(HasOperationalLimits):
+    """
+    Initialise the RequestWindowPluginActivity.
+
+    An `HasRequestWindowPluginActivity` instance initialises the waiting
+    on weather activity which is modelled using the
+    `RequestWindowPluginActivity`. For examples please refer to the
+    corresponding notebooks.
+
+    Parameters
+    ----------
+        offshore_environment: OffshoreEnvironment
+            An instance of the OffshoreEnvironment class (see below).
+    """
+
+    def __init__(
+        self, offshore_environment: "OffshoreEnvironment" = None, *args, **kwargs
+    ):
+        """Class constructor."""
+        super().__init__(*args, **kwargs)
+
+        if isinstance(self, model.PluginActivity) and isinstance(
+            offshore_environment, OffshoreEnvironment
+        ):
+            plugin = RequestWindowPluginActivity(
+                offshore_environment=offshore_environment
+            )
+            self.register_plugin(plugin=plugin, priority=2)
+
+
+# -------------------------------------------------------------------------------------!
+class RequestWindowPluginActivity(model.AbstractPluginClass):
+    """
+    Models the activity of a waiting on weather event.
+
+    The RequestWindowPluginActivity class is used to model the activity
+    of a waiting on weather event.
+
+    Parameters
+    ----------
+        offshore_environment: OffshorEnvironment
+            An instance of the OffshoreEnvironment class (see below).
+
+    """
+
+    def __init__(self, offshore_environment: "OffshoreEnvironment", *args, **kwargs):
+        """Class constructor."""
+        super().__init__(*args, **kwargs)
+
+        # Instance attributes
+        self.oe = offshore_environment
```

Comment:
It better to not use abriviations. Nobody is gonna remember that self.oe is offshore_environment. Just call it self.offshore_environment